src/lib/rule2F.ts

```diff
@@ -109,44 +187,80 @@ function getFunctionCalculatingAllowsNameFromContent(
 }
 
 /**
- * Calculates whether or not an element's name may be calculated using
- * its contents. Elements may allow name from content when they have certain
- * roles, be they explicit (role attribute) or implicit (semantic HTML).
- * @param elem - the function checks if 'elem' allows name from content
- * @return - true if elem allows name from content, false otherwise
+ * A container for lists of roles (explicit) and tags (implicit) that
+ * together account for a certain genre / type of role.
  */
-function allowsNameFromContent(elem: HTMLElement): boolean {
+interface RoleType {
+  roles: string[];
+  tags: string[];
+}
+
+/**
+ * Checks if a given HTMLElement matches any of the roles in a given RoleType.
+ * @param elem - element whose role type is in question.
+ * @param roleType - lists of indicators for some role type.
+ */
+function matchesRole(elem: HTMLElement, roleType: RoleType): boolean {
+  // Explicit roles : specified using 'role' attribute
   const explicitRole = elem.getAttribute('role')?.trim().toLowerCase() ?? '';
-  if (NAME_FROM_CONTENT_ROLES.includes(explicitRole)) {
+  if (roleType.roles.includes(explicitRole)) {
     return true;
   }
 
+  // Implicit roles : implied by current node tag name.
   const elemNodeName = elem.nodeName.toLowerCase();
-  if (NAME_FROM_CONTENT_ELEM_NODE_NAME.includes(elemNodeName)) {
+  if (roleType.tags.includes(elemNodeName)) {
     return true;
   }
 
-  const nameFromContentFunction = getFunctionCalculatingAllowsNameFromContent(
-    elemNodeName
-  );
-  if (nameFromContentFunction) {
-    return nameFromContentFunction(elem);
-  }
-
   return false;
 }
 
-export const TEST_ONLY = {allowsNameFromContent};
-
 /**
- * Checks if 'elem' in with 'context' satisfies the conditions
- * necessary to apply rule 2F.
+ * Checks if the contents of 'elem' with context 'context' should
+ * be used in its accesssible name. This is the condition for
+ * rule 2F.
  * @param elem - elem whose text alternative is being computed
  * @param context - additional information about the context of elem
  * @return - whether or not rule 2Fs condition has been satisfied
  */
-function rule2FCondition(elem: HTMLElement, context: Context): boolean {
-  if (allowsNameFromContent(elem)) {
+function allowsNameFromContent(elem: HTMLElement, context: Context): boolean {
+  // focasable elements should allow name from content.
```

Comment:
Should this check be combined with a check for membership of list 3?

> // As an override of list2, conditionally include name from content if current node is focusable, or if the current node matches list3 while the referenced parent node matches list1.

And it looks like the conditions for focusability  are more complicated than `hasAttribute('tabindex')`, see https://stackoverflow.com/questions/1599660/which-html-elements-can-receive-focus and https://html.spec.whatwg.org/multipage/interaction.html#focusable

Reply:
List 3 is handled lines 236-245.

I used the same stackoverflow page you linked as my source for checking whether an element is focusable, I think that 2F handles all of the examples given on this thread:
- `HTMLAnchorElement/HTMLAreaElement with an href` handled by 2F (172-179)
- `HTMLInputElement/HTMLSelectElement/HTMLTextAreaElement/HTMLButtonElement` Inputs, selects and textareas shouldn't take name from content according to the rest of the implementations (checked in the comparison app), as these are handled in special ways (2E). Buttons are handled in 2F line 34.
- `HTMLIFrameElement` Not sure what 'name from content' would look like for an iframe, in any case none of the other implementations allow name from content here.

Could be missing something here but hopefully that covers it :)